Ensure error handler does NEVER bubble up

kelunik opened the pull request:
Fixes #42.

kelunik committed:
Ensure error handler does NEVER bubble up

Fixes #42.

kelunik commented:
I have a further test case locally, but it doesn't work currently. It says the output is empty.

```
--TEST--
ErrorHandler::notify() fatals with a throwing error handler
--FILE--
<?php

require __DIR__ . "/../../vendor/autoload.php";

set_error_handler(function () {
    throw new RuntimeException;
});

AsyncInterop\Promise\ErrorHandler::notify(new Exception);

?>
--EXPECT--
Fatal error: Uncaught exception 'RuntimeException' while trying to report a throwing AsyncInterop\Promise::when() handler gracefully.
```

kelunik committed:
Skip exit test on PHP 5.4 and 5.5

PHPUnit doesn't support testing for STDERR on PHPUnit 4.x.

See https://github.com/sebastianbergmann/phpunit/commit/410644eb6a2e772c3b9be410cce1c7dee35a31b2.

kelunik committed:
Add error handling to README

jsor commented on composer.json:
```diff
@@ -7,7 +7,7 @@
     },
     "require-dev": {
-        "phpunit/phpunit": "^4|^5",
+        "phpunit/phpunit": "^4.8|^5.7",
```
Can you elaborate why exactly this is changed to that versions?

kelunik replied:
I'm not sure which one we actually need, for testing STDERR output, we need at least `5.1`. `4.0` does at not cover everything we need as Amp tests with `--prefer-lowest` have shown, so I just used the latest one here.

jsor replied:
So, probably change this to `"phpunit/phpunit": "^4|^5.1"`?

kelunik replied:
Any reason not to use the latest versions here? It's only a `require-dev` for this package, so compatibility with other constraints isn't required.

jsor replied:
Because that's an unrelated change (See also https://github.com/async-interop/promise/pull/43#issuecomment-274419903). You can still change the versions in a separate PR.

kelunik replied:
Well, it's not entirely unrelated, as we do require a newer version than was present for the new test. Anyway, I just fixed the skip test and now we require PHPUnit 4.8.13 anyway, see commit message for an explaination.

jsor commented on src/Promise/ErrorHandler.php:
```diff
@@ -79,19 +79,61 @@ public static function notify($error)
+        // The set error handler did throw or not exist, PHP's error handler threw, no chance to handle the error
+        // gracefully at this time. PANIC!
+
+        // Print error information to STDERR to the reason for the program abortion can be found, but do not expose
```
Not sure this makes sense.

kelunik replied:
Without it, you don't have any hint where it exited.

jsor replied:
Sorry, shoud have been more specific. I know what you're trying to say, but that sentence didn't parse over here (could be my fault though) :)

> Print error information to STDERR to the reason for the program abortion can be found

kelunik replied:
Indeed, that doesn't make sense. `to` → `so`.

jsor replied:
👍

jsor commented:
There are some unrelated changes in this patch, like namespacing functions and adding parenthesis to functions in the documentation.

I'd like to see changes like this handled in separate PRs in the future.

kelunik committed:
Fix typo

kelunik committed:
Fix skip statement for error handler test

We do not require a specific version of PHP, we only require PHPUnit 5.1. Previously, the test might have been run with PHP 7.0 and PHPUnit 4.8.x and then failed instead of being skipped.

PHPUnit 5.1 is required for the test to run successfully, PHPUnit 4.8.13 is required for PHPUnit_Runner_Version::series() and is included to run tests on PHP 5.5 and 5.4.

kelunik commented:
You're right, there are some unrelated changes. I usually try to keep them separate, but sometimes they're that minor that I don't think about it and fail to separate them.

kelunik committed:
Fix PHPUnit requirements

PHPUnit_Runner_Version::series() was present before, but has been fixed in https://github.com/sebastianbergmann/phpunit/commit/3c4becbce99732549949904c47b76ffe602a7595, which landed in the 4.8.25 and 5.3.3 release.

kelunik committed:
Fix skip section in error handler test 7

PHPUnit isn't loaded here, as it runs in a separate process. Failure to include the autoloader results in the test being executed.

kelunik commented:
I don't really like the `SKIPIF` test here. It (1) relies on PHPUnit being run from the directory and (2) relies on PHPUnit being run via the Composer installed version, but it could also be run using a PHAR in theory.

Does anybody have a better idea?

jsor commented on test/phpt/error_handler_007.phpt:
```diff
@@ -0,0 +1,31 @@
+require __DIR__ . "/vendor/autoload.php";
+
+$series = PHPUnit_Runner_Version::series();
+list($major, $minor) = explode(".", $series);
```
Is there a reason you don't just use `list($major, $minor) = explode(".", PHPUnit_Runner_Version::id(), 3);`?

kelunik replied:
Yes, that's fine, too. Doesn't really matter which one.

jsor replied:
Except that id() isn't broken in some versions of PHPUnit.

kelunik replied:
Sure, but with the current requirements, `::series()` isn't either. I can change it and adjust the requirement again, but I'd rather stop wasting time with changes that gain us literally nothing.

jsor commented on test/phpt/error_handler_007.phpt:
```diff
@@ -0,0 +1,31 @@
+ErrorHandler::notify() fatals with a throwing error handler
+--SKIPIF--
+<?php
+
```
Maybe something like this here:

```php
if (!file_exists(__DIR__ . "/vendor/autoload.php")) {
    die("Skipped: Cannot detect PHPUnit version, run tests from project root.");
}
```

kelunik committed:
Skip test when not able to detect vendor/autoload.php
